### PR TITLE
feat(eula): refactor eulas and add new config

### PR DESCRIPTION
**Pull request by Gankra**
This is an attempt to fix https://github.com/axodotdev/cargo-dist/issues/468

Apologies that it might be a bit over-engineered/complex, I got really lost in this EULA logic, especially with the multiple implementations of things kicking around.

Issues Being Addressed
----------------------

* There is no way to disable eula/license auto-detect
* There is no explicit Cargo.toml config for eula/license
  * Cargo's own license_file is weird, it's intended to be *instead of* license, for situations where your license is Too Weird For SPDX, so having our own dedicated field is desirable)
* The logic for licenses/eulas is vaguely duplicated, and requires several different locations to agree
* Using cargo-wix as a library, there's no way to know about Licenses that are supposed to be generated

Solutions
----------

Centralize License Logic
-------------------------

* Centralize all license/eula logic to eula.rs
* Do all license/eula logic all at once with a new Licenses type
* Replace the Eula enum with a License type that has more details resolved
* Add new license/eula configs to `[package.metadata.wix]`

Make wxs The License Generation Boss
------------------------------------

* Remove all eula/license handling from `initialize`, and make it the responsibility of `print::wxs`.
  * Rationale here is that it doesn't make a lot of sense to *only* produce the main.wxs when it has unresolved dependencies that were supposed to be generated.
  * With this approach if you want to know what the main.wxs should be, you also "have" to know about the other files that need to be generated. 
  * Caveat: `cargo wix print wxs` will still only print the one file, it probably should emit a warning that you have unresolved dependencies.
* Replace `wix::Execution::render_to_string` with `wix::Execution::render` which does all required rendering of `main.wxs/License.rtf` and returns several RenderOutputs ("the path this wants to be written to" plus "the string to write")
  * This is the interface that cargo-dist will want to use, so it can enumerate/diff/generate everything
* To avoid a bunch of gross complexities/generics, make the `wix::Execution::run` path that writes to disk/stdout just use `wix::Execution::render`
  * This means instead of writing straight to disk/stdout, we write to a String first
  * The overhead should be negligible given how small the files are, plus it makes it easier to test/debug stuff.
  * The alternative would be defining some kind of Output trait with a bunch of associated types and making all the code generic over it... no thanks.

Drive-by Changes
----------------

license-name is implemented but never used
------------------------------------------

Although I originally designed all the logic for setting/passing license-name, once I finished refactoring the code, there were no places that required it. Per your own comments in main.wxs.mustache:

> Change the value for the `Name` attribute in the `File` tag to the desired name for the file when it is installed alongside the `bin` folder in the installation directory. This can be omitted if the desired name is the same as the file name.

There are no places where we ever need the installed name to differ from the source name... kind of. There was one place in the code I found where the `Name` could be set to a non-trivial value, but it appears to be a typo:

https://github.com/volks73/cargo-wix/blob/81e49072f951477317fbf1583bd50e513679bba5/src/print/wxs.rs#L536

That code sets the name to `License` instead of `License.rtf` -- I'm not sure why you'd ever want the former. And indeed, I got a lot of confusing issues with different parts of the test suite seemingly disagreeing on that behaviour. Presumably this was the result of the license logic existing in several places? I normalized to `License.rtf`, which meant license-name was always None. Still, I left in the logic for license-name in case it ever becomes relevant again.

auto-licenses will be disabled if no output path is specified to write them to
----------------------------------------

I'm not sure if this is correct, but it was very confusing to do anything else. To make tests work I changed `print::wxs::Execution::default()` to set `output` to `wix\main.wxs`.

TODO
-------

* [x] write more doc-comments
* [x] add new config fields to docs
* [x] add tests for new config fields
* [x] restore some warnings that were lost in the refactor
* [x] fix tests (I think I lost some strip_prefix stuff)
* [x] test `cargo wix print wxs` stdout mode
   * [x] what should happen if there's dependent licenses???
* [ ] figure out if output path logic should be moved to print-wix as well
   * [ ] right now cargo-dist will need to manually say "output to wix\main.wxs" which is kinda whack but not exactly fatal
* [ ] do some integration-testing on cargo-dist to make sure this makes sense

### CODE REVIEW

**Comment by Gankra** [comment]
Although there are a couple details left to sort out, I'm marking this as ready for review because I'd like feedback on those details.

Things the subsequent commits do:

* Fixed the StoredPathBuf file_name parser to handle some more corner cases (tests added)
* Restored the strip_prefix behaviour to get relative paths
* Added back removed checks/diagnostics
* Added docs (with some driveby missing docs adds to things I noticed)
* removed wxs::Execution::default, replaced it with wxs::Execution::for_tests which takes the input path in (required for correctness)

**Comment by volks73** [comment]
Hello! Sorry for the delayed response. This is a very busy time of the year for me.

>There is no way to disable eula/license auto-detect

Interesting! This is an oversight in the API and feature set, or a long overdue TODO. It appears the current solution is to generate the WXS file and then remove the EULA UI element from within the XML. This is the same solution for removing the sidecar license file installed next to the binary. However, this works against creating ephermeal WXS files and/or configuration from the project's manifest.

There is the `-e,--eula` option for the `cargo wix init` and `cargo wix print` commands, but that is for the user to explicitly define a path to a RTF file instead of generating one from the project license. As noted, if it is not used, then the EULA is automatically generated. There is the `-l,--license` option which operates similar to the `-e,--eula` option, too.

It should be noted that there is a difference between an EULA and the license, both in practice and in the code. The EULA is not necessarily the same as the source code license. The EULA is what an end-user can do with the application, while the license is what a developer can or must do with the source code. A developer and end-user might be the same. Many projects share the license with the EULA. The actual correct usage is out-of-scope here, but I created this project, I did a high level survey of Rust projects and the license was typically used as an EULA. So, the default implementation was to use the license as the EULA in the MSI.

However, I apparently started to implement the ability to separate the EULA from the license with the `-e,--eula` and `-l,--license` options but never fully realized the feature.

I have to take a deep dive into the template and code, but I would think the EULA UI page of the MSI could be wrapped in an "optional" block and tied to a `--no-eula` flag and a similar `--no-license` flag for the sidecar file.

> There is no explicit Cargo.toml config for eula/license
>    - Cargo's own license_file is weird, it's intended to be instead of license, for situations where your license is Too Weird For SPDX, so having our own dedicated field is desirable)

Yes, the weirdness with the SPDX and having custom licenses was the original motivator for the `-e,--eula` and `-l,--license` options.

>The logic for licenses/eulas is vaguely duplicated, and requires several different locations to agree

This is probably due to the half-implemented separation of between EULAs and licenses.

All-in-all, it looks like the EULA and License management features could use some more attention and refinement. I am open merging any PRs related to this work.

File: src/eula.rs [hunk 1]
```diff
@@ -1,116 +1,272 @@
...
+use cargo_metadata::Package;
+
+/// License info
+#[derive(Clone, Debug)]
+pub struct Licenses {
+    /// The license for the actual source code
+    ///
+    /// This likely will become/contain a Vec at some point,
+    /// since dual MIT/Apache wants to have two license files!
+    pub source_license: Option<License>,
```
Comment (volks73) [inline src/eula.rs:39]: Can this just be called `source` and `end_user`? We already know these are licenses based on the type definition and being fields in the `Licenses` struct.

File: src/eula.rs
Comment (volks73) [inline src/eula.rs:1]: Should this file be renamed to `licenses.rs` instead of `eula.rs`? The module is for all licenses, not just the EULA.

File: src/eula.rs [hunk 2]
```diff
@@ -1,116 +1,272 @@
...
+        trace!("finding source license for {}", package.name);
+        // If explicitly passed, use that
+        if let Some(path) = path {
+            trace!("explicit source-license path passed as argument, using that");
+            return Ok(Some(License::from_stored_path(path)));
+        }
+
+        let package_dir = package.manifest_path.parent().expect("non-root Cargo.toml");
+
+        // If there's [package.manifest.wix].license, respect that
```
Comment (volks73) [inline src/eula.rs:112]: This is conflating the EULA and the sidecar License file, correct? Should there ultimately be a `eula` and `license` field? Maybe we actually need a hierarchy of some kind, not to over complicate, but a `license` field is the parent and if a `eula` and `sidecar` (or whatever) are defined, they override the `license` field, but if not, then the license is used for both the `eula` and `sidecar` and if the `license` is not defined, then auto-generated is used. Not sure about disabling one or both with a single field.

Comment (volks73) [reply src/eula.rs:112]: Reading further down, there does appear to be a `eula` field separate from `license`, but I am still wondering about the hierarchy.

File: src/lib.rs [hunk 3]
```diff
@@ -1159,25 +1151,51 @@ impl StoredPath {
 
     /// Returns the final component of the path, if there is one.
     pub fn file_name(&self) -> Option<&str> {
+        let mut path = self.as_str();
```
Comment (volks73) [inline src/lib.rs:1154]: Sorry, I am coming into this a little cold. Why is a custom implementation of finding a file name needed? All of the functionality around `StoredPath` really feels like it should its own crate/package or might already be provided by an existing crate/package. I just know there be dragons here.

Comment (volks73) [reply src/lib.rs:1154]: Maybe to start we can move all of the stored path code to a separate module from the `lib.rs` module?

File: src/main.rs [hunk 4]
```diff
@@ -786,9 +790,12 @@
 //! Available for the _init_ (`cargo wix init`) and _print_ (`cargo wix print`)
 //! subcommands.
 //!
-//! Sets the path to a bitmap (.bmp) image file that will be displayed across
+//! Sets the path to an image file (.png, .bmp, ...) that will be displayed across
```
Comment (volks73) [inline src/main.rs:793]: Are we sure a PNG can be used? Maybe PNG support was added to WiX after I originally implemented this feature, but I vaguely remember only bitmaps being supported.

File: src/print/mod.rs [hunk 5]
```diff
@@ -33,6 +34,45 @@ use std::path::PathBuf;
 
 use cargo_metadata::Package;
 
+/// The result of rendering a template (main.wxs, License.rtf, ...)
+pub struct RenderOutput {
```
Comment (volks73) [inline src/print/mod.rs:38]: Should this eventually implement the `std::io::Write` trait?

We want to write either to a String, STDOUT, or a File for most output. The `std::io::Cursor`could be used for writing to a String, the `std::io` "stuff" could be used for STDOUT or a File. Again, coming in cold and trying to wrap my brain around this.

**Review by volks73 — COMMENTED** [review]
Thank you for the PR and the work. I like the direction this is going, and I appreciate fixing the documentation, tests, and small improvements. As mentioned, I am a little distracted at the moment, so I apologize for the fly by review and I am trying to get by mind wrapped around the changes.

File: src/main.rs (hunk 4 above)
Comment (Gankra) [reply src/main.rs:793]: Yeah even WiX's docs don't say it's allowed but I tested it out and png's work great (and should probably be preferred for obvious reasons).

File: src/eula.rs
Comment (Gankra) [reply src/eula.rs:1]: Yeah I had considered that but wanted to avoid kicking over your whole sand castle here. But since you agree, happy to do it :)

File: src/eula.rs (hunk 2 above)
Comment (Gankra) [reply src/eula.rs:112]: The current hierarchy is that 

* we try all the different ways to find a source-license
* we try all the different ways to find an end-user-license
  * the last option being "use the source-license if it's an RTF (and eula=false wasn't set)"
  
I think it makes sense that they should be treated as separate/equals (as the code currently does). Personally I think the "use your source-license as a eula" thing is a bit of a silly cargo-cult, but as you say, people like doing it, so I'm happy to keep it working for people who like it (that said, cargo-dist will probably default to emitting eula=false/license=false into the user's Cargo.toml to make this stuff opt-in instead of opt-out).

File: src/lib.rs (hunk 3 above)
Comment (Gankra) [reply src/lib.rs:1154]: I agree it's increasingly wanting to be a separate crate -- the reason a separate impl is needed is because std::path/camino both have host-platform-specific behaviours on paths, and there's no way to ask for anything else. 

Most notably, `\` is a valid character for a unix filename, so if you're on linux `a/b\c.txt` will interpretted as `filename = "b\c.txt"`, which is not desirable for us.

File: src/print/mod.rs (hunk 5 above)
Comment (Gankra) [reply src/print/mod.rs:38]: Hmm not quite sure what you're trying to describe -- the RenderOutput type is really intended to just be a String, but I wanted to get rid of a place where two parts of the code have to assume the location they're talking about.

Basically when rendering out a template we really need to know where on disk that file will be written, because templates contain relative paths. So since the template renderer already has to know the final path, I decided it was best for it to report back that path so that the caller didn't have to assume/know anything, and could just faithfully use the reported path.

And then once it already stored the path I was like "well ok I can just have a method to write it to disk". There is however an outstanding issue -- as designed, the "print to stdout" mode definitionally has no know target path, and can only output one file (when we may need to output up to 3 (well really only 2))... not sure how to make that work.

Comment (volks73) [reply src/print/mod.rs:38]: I was more or less just "thinking out loud" while I was reviewing the code. I was wondering if the `std::io` already had some functionality that could be reused but I was misremembering from when I was trying to have output be either STDOUT or a file and there were some traits I had to play around with to make it all work.

**Comment by Gankra** [comment]
Did all the discussed refactors (just made stored_path a module, not a crate).

@volks73 how would you feel about disabling the "stdout" mode of `cargo wix print wxs` (effectively making the output path a required argument). I'm really struggling to rationalize how that's supposed to work correctly with us generating licenses and needing to emit relative paths to things.

(Current status of the PR has it as "still a thing you can ask for, but the output is totally broken")

**Comment by volks73** [comment]
>I'm really struggling to rationalize how that's supposed to work correctly with us generating licenses and needing to emit relative paths to things.

@Gankra The `cargo wix print wxs` should not be generating any other output other than a WXS, a.k.a. XML, file. I am missing the issue associated with printing to STDOUT. Requiring the output argument be a path to a file would be the equivalent to redirecting output to file, correct? E.g., `cargo wix print wxs > somefile.wxs` is equivalent to `cargo wix print wxs somefile.wxs`. But, I admit I have still not done a detailed review of this PR yet.

The `print` subcommand generates a string. If the paths are relative it is relative to the current working directory when the user issued the command. Should absolute paths be used instead, or the option/flag for absolute paths?

I was thinking the `cargo wix print wxs | cargo wix` workflow was going to be "a thing" for CI and ephemeral builds (https://github.com/volks73/cargo-wix/issues/88), i.e. no artifacts or WXS file would be needed to be included in the source code. Reading a WXS file from STDIN is currently not implemented for the `cargo wix` command, but it was something I was thinking about as a feature.

My original thinking was that: `cargo wix init` is equivalent to:

```sh
cargo wix print MIT > ./wix/license.rtf
cargo wix print WXS > ./wix/main.wxs
```

Would a `cargo wix print --working-dir=./example/path WXS` be useful and resolve the relative path issue? All paths in the printed template to STDOUT would be relative to the working directory that was passed. My desire for consistency and symmetry is kicking in with having the WXS template be "special" and require an argument while others do not.

**Comment by Gankra** [comment]
Thanks for the reply, I think that shook loose a weird assumption I had stuck in my craw (that the paths should be output-dir-relative, rather than purely input-dir-relative). I think I can finish this quickly now.

**Comment by Gankra** [comment]
@volks73 i believe this PR is complete/landable now. cargo-dist can now generate all the subfiles and stuff for people who want the eula functionality:

image

**Comment by Gankra** [comment]
(got the downstream patch ready, in case you're curious to see what library usage of this API is like: https://github.com/axodotdev/cargo-dist/pull/503)

**Review by volks73 — APPROVED** [review]
Looks good.
